packages/react/src/components/List/List.tsx

```diff
@@ -0,0 +1,77 @@
+import { useMergeRefs } from '@floating-ui/react';
+import { Slot } from '@radix-ui/react-slot';
+import cl from 'clsx/lite';
+import { forwardRef, useEffect, useId, useRef } from 'react';
+
+import { Paragraph } from '../Typography';
+
+export type ListProps = {
+  /**
+   * Change the default rendered element for the one passed as a child, merging their props and behavior.
+   * @default false
+   */
+  asChild?: boolean;
+  /**
+   * Changes text sizing
+   * @default md
+   *
+   */
+  size?: 'sm' | 'md' | 'lg';
+  /**
+   * Changes type of list style
+   * @default unordered
+   *
+   */
+  variant?: 'unordered' | 'ordered';
+} & Omit<
+  | ({ variant: 'ordered' } & React.OlHTMLAttributes<HTMLOListElement>)
+  | ({ variant?: 'unordered' } & React.HTMLAttributes<HTMLUListElement>),
+  'size'
+>;
+
+export const List = forwardRef<HTMLOListElement, ListProps>(function List(
+  {
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledby,
+    asChild,
+    className,
+    size = 'md',
+    variant = 'unordered',
+    ...rest
+  },
+  ref,
+) {
+  const Component = asChild ? Slot : variant === 'ordered' ? 'ol' : 'ul';
+  const randomId = useId();
+  const innerRef = useRef<HTMLOListElement>(null);
+  const hasLabel = ariaLabel ?? ariaLabelledby ?? false;
+  const mergedRefs = useMergeRefs([innerRef, ref]);
+
+  // Connect previous heading element if available. This does not affect visuall rendering,
+  // and is not an accessibility requirement, but its a nice enhancement when heading + list relation is obvious.
+  useEffect(() => {
+    const list = innerRef.current;
+    const heading = list?.previousElementSibling;
+    const headingId = heading?.id || randomId;
+
+    if (hasLabel === false && heading?.matches('h1,h2,h3,h4,h5,h6')) {
+      list?.setAttribute('aria-labelledby', headingId);
+      heading.id = headingId;
+    }
+    return () => list?.removeAttribute('aria-labelledby');
+  }, [hasLabel]);
```

Comment:
Doing this feels a bit weird.
This might be feel like magic to the user, and is something they **need to know** when using the list, compared to this api:
```jsx
<List.Provider>
  <List.Heading>Heading</List.Heading>
  <List.Ordered>
    <List.Item>item</List.Item>
  </List.Ordered>
</List.Provider>
```

I think we should drop doing hidden things to headings before the list in the dom, as this also opens up for edge cases where this breaks. I suggest going for a provider, that does not render a div, and explicitly have a `List.Heading` element.

There _might_ be cases where the heading before the list is not at all connected to the list, and this solution does not give an option to opt out,

Reply:
> Det ville vært et svært dårlig design 
Gjerne utdjup på kvifor dette er dårlig?

Reply:
> There might be cases where the heading before the list is not at all connected to the list, and this solution does not give an option to opt out,
Å ha en overskrift rett før en liste, som ikke er overskriften til listen? Det vil være et leit design fordi det da visuelt indikerer en sammenheng som ikke egentlig skal være der :)